mocket/recording.py

```diff
@@ -0,0 +1,170 @@
+from __future__ import annotations
+
+import binascii
+import contextlib
+import hashlib
+import json
+from collections import defaultdict
+from dataclasses import dataclass
+from pathlib import Path
+
+from mocket.compat import decode_from_bytes, encode_to_bytes
+from mocket.types import Address
+
+hash_function = hashlib.md5
+
+with contextlib.suppress(ImportError):
+    from xxhash_cffi import xxh32 as xxhash_cffi_xxh32
+
+    hash_function = xxhash_cffi_xxh32
+
+with contextlib.suppress(ImportError):
+    from xxhash import xxh32 as xxhash_xxh32
+
+    hash_function = xxhash_xxh32
+
+
+def _hash_prepare_request(data: bytes) -> bytes:
+    _data = decode_from_bytes(data)
+    return encode_to_bytes("".join(sorted(_data.split("\r\n"))))
+
+
+def _hash_request(data: bytes) -> str:
+    _data = _hash_prepare_request(data)
+    return hash_function(_data).hexdigest()
+
+
+def _hash_request_fallback(data: bytes) -> str:
+    _data = _hash_prepare_request(data)
+    return hashlib.md5(_data).hexdigest()
+
+
+def hexdump(binary_string: bytes) -> str:
+    r"""
+    >>> hexdump(b"bar foobar foo") == decode_from_bytes(encode_to_bytes("62 61 72 20 66 6F 6F 62 61 72 20 66 6F 6F"))
+    True
+    """
+    bs = decode_from_bytes(binascii.hexlify(binary_string).upper())
+    return " ".join(a + b for a, b in zip(bs[::2], bs[1::2]))
+
+
+def hexload(string: str) -> bytes:
+    r"""
+    >>> hexload("62 61 72 20 66 6F 6F 62 61 72 20 66 6F 6F") == encode_to_bytes("bar foobar foo")
+    True
+    """
+    string_no_spaces = "".join(string.split())
+    return encode_to_bytes(binascii.unhexlify(string_no_spaces))
```

Comment:
I'd leave those functions under utils. They are not specific to the "recording" feature, even though only used it for now.

Reply:
sure can do.
i personally always prefer keeping code as tightly scoped as possible to prevent the start of code smells :P

Reply:
moving those functions back to utils, the `_load` function needs to handle `binascii.Error` tho. (which kinda relies on knowledge of the implementation of `hexload`.

should i replace that with a generic `ValueError` to decouple ?

Reply:
I understand it, but it's probably something people will be willing to change, when inheriting from the storage class. There are multiple reasons for me to have chosen to hex dump, but I reckon users may have different needs.

Reply:
rrr. now i remember the other reason i moved the functions.

because of the backwards-compat re-exports of `MocketSocketCore` and `MocketMode` in `utils.py` having `hexdump` and `hexload` in there, now causes a cyclic-import :(

Reply:
so yeah. without removing those backwards-compat re-exports i cannot move the functions to utils :(

Reply:
Makes sense, I did not think about that.

Reply:
so what do you prefer. remove backwards-compat re-exports or keep hexdump/hexload in `recording.py` ?

Reply:
Let's get rid of the import lines and move the functions back to `utils`.

Reply:
ok. will do

Reply:
got an opinion on the above question regarding `binascii.Error` vs `ValueError` ?

Reply:
Sorry, I completely missed it. I like the idea of using a `ValueError`.

Reply:
ok will do.

Reply:
done